tools/benchmark.sh

```diff
@@ -0,0 +1,172 @@
+#!/bin/bash
+set -o errexit
+set -o pipefail
+set -o nounset
+#set -o noclobber
+
+REPO=$(git rev-parse --show-toplevel)
+
+RED="\033[31m"
+GREEN="\033[32m"
+BLUE="\033[34m"
+RESET="\033[0m"
+
+## Parallel arrays because an array of tuples is too much to ask for in Bash
+BENCHMARK_NAMES=(
+    "herostratus"
+    "git"
+)
+BENCHMARK_URLS=(
+    "git@github.com:Notgnoshi/herostratus.git"
+    "https://github.com/git/git.git"
+)
+BENCHMARK_BRANCHES=(
+    "main"
+    "master"
+)
+
+debug() {
+    echo -e "${BLUE}DEBUG:${RESET} $*" >&2
+}
+
+info() {
+    echo -e "${GREEN}INFO:${RESET} $*" >&2
+}
+
+error() {
+    echo -e "${RED}ERROR:${RESET} $*" >&2
+}
+
+usage() {
+    echo "Usage: $0 [--help]"
+    echo
+    echo "Benchmark Herostratus on a few different popular repositories"
+    echo
+    echo "  --help, -h      Show this help and exit"
+    echo "  --no-edit, -n   Do not edit the README with benchmark results"
+}
+
+## Execute a command, and measure its wall-clock execution time in seconds
+#
+# Does not capture command's stderr/stdout output. Saves the execution time to the global variable
+# EXEC_RUNTIME.
+EXEC_RUNTIME="0"
+exec_with_timing() {
+    debug "Executing: $*"
+    local -r start=$(date +%s.%2N)
+    "$@"
+    local -r end=$(date +%s.%2N)
+    EXEC_RUNTIME=$(awk "BEGIN {print $end - $start}")
+}
+
+## Run a single benchmark on the given repository
+#
+## Outputs the results to stdout as a single row in a markdown table.
+run_benchmark() {
+    local -r data_dir="$1"
+    local -r name="$2"
+    local -r url="$3"
+    local -r branch="$4"
+
+    exec_with_timing herostratus --data-dir "$data_dir" add --name "$name" "$url" "$branch" >&2
+    local -r add_time="$EXEC_RUNTIME"
+
+    local -r stdout="$data_dir/check-$name.out"
+    local -r stderr="$data_dir/check-$name.err"
+
+    exec_with_timing herostratus --data-dir "$data_dir" --color check-all --no-fetch \
+        >"$stdout" \
+        2> >(tee "$stderr" >&2)
+    local -r check_time="$EXEC_RUNTIME"
+
+    local num_achievements
+    num_achievements="$(grep "Generated.*achievements" "$stderr" | sed -En 's/^.*Generated ([0-9]+) achievements.*$/\1/p')"
+
+    local num_commits
+    num_commits="$(grep "Generated.*achievements" "$stderr" | sed -En 's/^.*processing ([0-9]+) commits.*$/\1/p')"
+    echo "| $name | $branch | $num_commits | ${add_time}s | ${check_time}s | $num_achievements |"
```

Comment:
As much fun as this was, I think `herostratus check` and `check-all` should have a `--summary` flag to output this table themselves.

* [ ] This script should pass `--no-fetch` to `add` and have `check-all` output the summary with the fetch time and processing time
* [ ] It should also include time spent per-commit.